apps/meteor/ee/tests/unit/server/hooks/messages/BeforeSaveCannedResponse.tests.ts

```diff
@@ -0,0 +1,169 @@
+import { registerModel } from '@rocket.chat/models';
+import { expect } from 'chai';
+import { before, describe, it } from 'mocha';
+import sinon from 'sinon';
+
+import { BaseRaw } from '../../../../../../server/models/raw/BaseRaw';
+import { BeforeSaveCannedResponse } from '../../../../../server/hooks/messages/BeforeSaveCannedResponse';
+
+const createMessage = (msg?: string, extra: any = {}) => ({
+	_id: 'msg-id',
+	rid: 'GENERAL',
+	ts: new Date(),
+	u: {
+		_id: 'user-id',
+		username: 'user',
+	},
+	_updatedAt: new Date(),
+	msg: msg as string,
+	...extra,
+});
+
+const createRoom = (extra: any = {}) => ({
+	_id: 'GENERAL',
+	name: 'general',
+	...extra,
+});
+
+const createUser = (extra: any = {}) => ({
+	_id: 'user-id',
+	name: 'User Name',
+	username: 'user',
+	emails: [{ address: 'user@user.com' }],
+	...extra,
+});
+
+class LivechatVisitorsModel extends BaseRaw<any> {
+	findOneEnabledById() {
+		return {};
+	}
+}
+
+class UsersModel extends BaseRaw<any> {
+	async findOneById() {
+		return {
+			name: 'John Doe Agent',
+		};
+	}
+}
+
+const db = {
+	collection: () => ({}),
+};
+
+describe('Omnichannel canned responses', () => {
+	before(() => {
+		registerModel('ILivechatVisitorsModel', () => new LivechatVisitorsModel(db as unknown as any, 'visitor'));
+		registerModel('IUsersModel', () => new UsersModel(db as unknown as any, 'user'));
+	});
+
+	it('should do nothing if canned response is disabled', async () => {
+		BeforeSaveCannedResponse.enabled = false;
+
+		const canned = new BeforeSaveCannedResponse();
+
+		const message = await canned.replacePlaceholders({
+			message: createMessage('{{agent.name}}'),
+			room: createRoom({ t: 'l', servedBy: { _id: 'agent' }, v: { _id: 'visitor' } }),
+			user: createUser(),
+		});
+
+		return expect(message).to.have.property('msg', '{{agent.name}}');
+	});
+
+	it('should do nothing if not an omnichannel room', async () => {
+		BeforeSaveCannedResponse.enabled = true;
+
+		const canned = new BeforeSaveCannedResponse();
+
+		const message = await canned.replacePlaceholders({
+			message: createMessage('{{agent.name}}'),
+			room: createRoom(),
+			user: createUser(),
+		});
+
+		return expect(message).to.have.property('msg', '{{agent.name}}');
+	});
+
+	it('should do nothing if the message is from a visitor', async () => {
```

Comment:
this is a new behavior I introduced but think makes sense.. wdyt @KevLehman ?

Reply:
Makes sense, we don't allow canned responses from visitors 👀

Reply:
it is weird though, the implementation replaces the placeholder for every message, not only the ones where canned responses were used.. 🤷

Reply:
Yeah, that's because we don't have a way to differentiate if a canned response was used or not in a message, since the message comes from the client directly (as a normal message) :sadblob: